text_recognizer/data/iam_original_and_synthetic_paragraphs.py

```diff
@@ -39,9 +39,22 @@ def setup(self, stage: str = None) -> None:
         if stage == "fit" or stage is None:
             self.data_train = ConcatDataset([self.iam_paragraphs.data_train, self.iam_syn_paragraphs.data_train])
             self.data_val = self.iam_paragraphs.data_val
+
         if stage == "test" or stage is None:
             self.data_test = self.iam_paragraphs.data_test
 
+    def train_dataloader(self):
```

Comment:
I think putting the data generation inside this method is a neat way to quickly get "infinite data" with Lightning. But I think we need to be careful not to break assumptions/promises of the Lightning API.

1. We're calling the `.setup` method -- that's part of the Lightning API, with expectations and guarantees about when it's called. We don't want a future contributor (e.g. us in six months) to expect it to conform to that API and end up burned. I think this is easily resolvable though. Elsewhere, I suggested refactoring out `_load_dataset` from `.setup` and making it its own method. We could call that method both here and in `.setup`.

2. For the first epoch and in the case where we never do a dataloader reload, it seems like we're needlessly generating the data twice: first when `setup` gets called and then again when `train_dataloader` gets called. It should only be generated once. That should be done in `.setup`, since it's expected that `.setup` does everything that's needed to prep a dataset. We could set a flag when `.setup` gets called 

The other way to do "infinite data" would be to incorporate the image synthesis into the `Dataset` itself. This would be a new class that lives in `iam_synthetic_paragraphs`, inheriting from the `torch.utils.data.Dataset`.  You'd use the `__getitem__` method to call something like `generate_synthetic_paragraphs`, but working one paragraph at a time. You'll want to handle `transforms` the same way as `BaseDataset`.

There's a potential for slowdown, but I suspect our batch size limitations from GPU RAM are enough that multiprocessing `DataLoader`s can generally serve up a full batch by the time the GPU finishes a fwd/bwd pass. You can use the `--profile` flag (with `--gpus=1, `--max_epochs=1` and the same `--num_workers` as in the `ddp` case) to check the detailed trace or just use the GPU utilization reported to W&B by `nvidia-smi` as a spot check (should be 70%+). 

See StackOverflow discussion of infinite datasets in PyTorch [here](https://stackoverflow.com/questions/54359243/implementing-an-infinite-loop-dataset-dataloader-in-pytorch) and watch out for the RNG issues they mention!

Note that `__len__` can be set to return whatever number you wish as needed, e.g. by a command line argument. 

FYI for 2, I think the pseudocode version of what Lightning is doing with the `DataModule` is something like:
```python
# download, etc...
dm = TheDataModule()
dm.prepare_data()

# splits/transforms
dm.setup(stage="fit")

if not reload_dataloaders_every_n_epochs:
    train_loader = dm.train_dataloader()
# fitting loop
for epoch in epochs:
  if reload_dataloaders_every_n_epochs and not epoch % reload_dataloaders_every_n_epochs:
        train_loader = dm.train_dataloader()

  for batch in dm.train_dataloader():
      ...
  
  for batch in dm.val_dataloader():
      ...

dm.teardown(stage="fit")

Reply:
Check out https://github.com/full-stack-deep-learning/fsdl-text-recognizer-2022/pull/40